### ISSUE TITLE
Provide better keybindings for the embark integration

### ISSUE THREAD

**Issue by lhernanz** [armindarvish/consult-gh#174]
The current bindings for consult-gh-embark are using multiple prefix keys with only a small number of commands in each. That implies that you need to remember around 10 different prefixes without any visual help from embark that just shows a `prefix` keyword associated to the top level key. To make the situation even more challenging, there is also a couple of second level prefixes that just multiply the problem. I have also been unable to find any easy way to relate the prefixes to the different commands that they execute. Overall, this makes the integration very challenging to use. 

I would really appreciate it if you could figure out different bindings that are more straightforward to use. It will be challenging, as there are a ton of good commands, but my recommendation would be:

- Use less prefixes overall, try to avoid second level prefixes
- Try to figure out an easy to remember link between prefixes and commands. Some ideas might be:
  -  Group them relative to the top object. All the actions related to a repo could be under "r" (now they are split between br, f, and r). This would probably be my preferred choice as it seems that you only have a few high level objects in the common map: repo, org, links, conf?
  - All commands starting by the same letter could use that prefix (now c commands are split between c, l, r, and w)
- When possible, don't repeat the commands. It just adds cognitive load
- Try to shorten the names as Embark is not displaying the full name sometimes. This is tricky due to the long consult-gh-embark prefix.

Thanks

**Comment by armindarvish** [comment]
@lhernanz Thanks for the detailed comment, but unfortunately there is never going to be a perfect set of keybindings that fits everyone's workflow. You are very much supposed to change those key bindings to fit your own work flow. This is relatively easy since the bindings are in their own keymaps (e.g. `consult-gh-embark-general-actions-map`, `consult-gh-embark-issues-actions-map
=, =consult-gh-embark-prs-actions-map`, etc.)  

Now to address some of the points in your comments for the sake of discussion (for anyone who might be interested in the topic):  

> -   Use less prefixes overall, try to avoid second level prefixes.

The multi-level prefixes is because, consult-gh tries to respect default embark bindings.  

> -   Try to figure out an easy to remember link between prefixes and commands. Some ideas might be:  
>     -   Group them relative to the top object. All the actions related to a repo could be under "r" (now they are split between br, f, and r). This would probably be my preferred choice as it seems that you only have a few high level objects in the common map: repo, org, links, conf?
>     -   All commands starting by the same letter could use that prefix (now c commands are split between c, l, r, and w)

If you upgrade to the latest version, the first point above is somewhat addressed. Everything about a repo is under "r" for example. In other cases, consult-gh tries to use familiar embark or emacs bindings, for example `i` for insert, `w` for kill to ring, `f f` for find files, &#x2026; There is a logic behind some of these bindings even if it is not obvious at first sight. Is it perfect? Of course not, but there is no such a thing as a perfect key binding for every one and every scenario to begin with.  

> -   When possible, don't repeat the commands. It just adds cognitive load

I disagree here, the fact that there are multiple paths to the same destination does not mean that one needs to know every one of them and actively choose between them. One can take whatever path works for them and forget about the rest. Of course some users may have an obsession with optimization which may lead to an endless search for better workflows, which can cause quite a lot of cognitive load. This is often common in Emacs world, because the kind of people taht choose to use Emacs are often the kind of people who have a tendency to want to optimize everything down to every single key press! Such users will likely suffer from the cognitive load you mention regardless of what Emcs and/or packages provide, but that needs to be addressed on the suer side not on the tool side!  

> -   Try to shorten the names as Embark is not displaying the full name sometimes. This is tricky due to the long consult-gh-embark prefix.

This depends on your Embark config. You can set things up so that you can use `C-h` to get all commands with searchable completion like the screenshot below:  

Overall, I think there are enough possibilities for any user to build their own workflow with keybindings they like and in the manner they like. This is Emacs world after all not "vs code"!

**Comment by lhernanz** [comment]
Thanks for the detailed explanation. I was not aware about the "C-h" trick that is useful in this particular case. 

But, as you are showing in your screenshot, I am also using the which-key indicator and the "+prefix" entries were bothering me as well as the long function names that don't fit into the space that which-key is allocating to them. Therefore I have implemented a couple of changes in my conf to solve both problems. 

I am going to share them here in case that they are useful for anybody else:

- I am creating empty maps to provide names to the prefixes
- I have created a function to shorten the function names by removing the package names from them

I prefer the new user experience after these changes.

```elisp
(defun my-short-def (function)
    "Return a shortened name definition of the FUNCTION to be used in a keymap.
The format is (STRING . DEFN) as defined in `keymap-set'. The
shorter name is convenient for displaying the keymap using
which-key that does not reserve a lot characters for every
function name. The short name is computed by removing the package
name from the function name."
    (let* ((func-name (symbol-name function))
           (package-file (symbol-file function 'defun))
           (package-name (file-name-sans-extension (file-name-nondirectory package-file))))
      (cons (string-remove-prefix (concat package-name "-") func-name) function)))
```

```elisp
  (defvar-keymap consult-gh-embark-general-actions-map
      :doc "Keymap for consult-gh-embark"
      :parent embark-general-map
      "b" `("gh-orgs" . ,(make-sparse-keymap))
      "b o" (my-short-def #'consult-gh-embark-add-org-to-known-orgs)
      "b k" (my-short-def #'consult-gh-embark-remove-org-from-known-orgs)
      "b d" (my-short-def  #'consult-gh-embark-add-org-to-favorite-list)
      "b D" (my-short-def #'consult-gh-embark-remove-org-from-favorite-list)
      "c" `("gh-create" . ,(make-sparse-keymap))
      "c i" (my-short-def #'consult-gh-embark-create-issue)
      "c p" (my-short-def #'consult-gh-embark-create-pr)
  ...)
```

<img width="940" alt="image" src="https://github.com/user-attachments/assets/7f691631-72e5-4706-bb1c-31eff06f84cc" />

<img width="493" alt="image" src="https://github.com/user-attachments/assets/f0d5ef2f-1a7d-4880-9a41-76f47af92842" />

**Comment by armindarvish** [comment]
@lhernanz if the main issue is the name in the `which-key` menu, an alternative and simpler  solution is to remove the `"consult-gh-embark-"` in `which-key-replacement-alist`:  

```emacs-lisp
(add-to-list 'which-key-replacement-alist '((nil . "consult-gh-embark-") . (nil . "")))
```

**Comment by lhernanz** [comment]
No, the main issue is to avoid the name-less keymaps that make the main screen of embark not very user-friendly for me. See screenshot below vs the one above with the changes that I have using now. You told me about the `C-h` trick that, for sure, it is a mitigation for this problem, but I am more used to use which-key with the shortcuts.

<img width="960" alt="image" src="https://github.com/user-attachments/assets/3fc4fc6d-fd83-4c26-9fe4-26bd4fa765d0" />

I did not know about `which-key-replacement-alist`! Agree that it is a better way to shorten the name. Thanks for the tip.

Exploring deeper, I see that you also can name the prefixes also using the following which-key feature:
```elisp
(which-key-add-keymap-based-replacements  consult-gh-embark-general-actions-map "r" "gh-repos")
```

**Comment by armindarvish** [comment]
@lhernanz I just improved the embark action and menus in a new branch:  
[consult-1.10](https://github.com/armindarvish/consult-gh/tree/consult-1.10)  

Note that that branch only works with the latest updates in consult (See [ada079d](https://github.com/minad/consult/commit/ada079d5932700a8819ace622ef4323e73983161)), but you can check out the embark actions and see if it's better for what you wanted.  

For context, the solution with `which-key-add-keymap-based-replacements` is buggy becuase it adds some noise to the `C-h` functionality. Specifically, if you press `C-h`, you would see options for selecting prefixes that causes an error, but with the new approach I just pushed in the new branch, everything works as expected. You get menu names for prefixes in the which-key menu and `C-h` works as expected too.  

I also added a bunch of new embark actions that should make finding things easier.

**Comment by armindarvish** [comment]
@lhernanz I have merged the embark improvements in the consult-1.10 branch with develop and main. So with the latest consult-gh on MELPA (or manually installing from main or develop branch) you can now get the improved embark keybinding that looks better in which-key menu as well.
